File: java/GenerateParentheses.java
public class Solution {
    public int n;
    public List<String> result = new ArrayList<String>();
    public List<String> generateParenthesis(int n) {
        this.n = n;
        StringBuilder valuelist = new StringBuilder();
        this.dfs(0, 0, valuelist);
        return this.result;
    }
    
    public void dfs(int left, int right, StringBuilder valuelist) {
        if (left == this.n && right == this.n) {
            this.result.add(valuelist.toString());
            return;
        }
        if (right > left || left > this.n) {
            return;
        }
        if (left > right) {
            StringBuilder newList = new StringBuilder(valuelist);
            newList.append(')');
            this.dfs(left, right+1, newList);
        }
        StringBuilder newList = new StringBuilder(valuelist);
        newList.append('(');
        this.dfs(left+1, right, newList);
    }
}

File: archive/python/backtracking/GenerateParentheses.py
class Solution:
    # @param an integer
    # @return a list of string
    def __init__(self):
        self.result = []
        self.n = 0
    
    def dfs(self, l, r, s):
        if l == r and r == self.n:
            self.result.append(s)
            return
        if l < r:
            return
        if l <= self.n:
            self.dfs(l+1, r, s+"(")
        self.dfs(l, r+1, s+")")
        
    def generateParenthesis(self, n):
        self.n = n
        self.dfs(0, 0, "")
        return self.result
        
if __name__ == "__main__":
    solution = Solution()

    print solution.generateParenthesis(3)
